### ISSUE TITLE
Child Window Drawing Order

### ISSUE THREAD

**Issue by ElectroidDes** [ocornut/imgui#7964]
### Version/Branch of Dear ImGui:

Version 1.90.8

### Back-ends:

OpenGl

### Compiler, OS:

Windows10

### Full config/build information:

Why does the child window "CHILD_2", which, as can be seen from the code, is called last after all the other windows, draw as if it is BETWEEN the window "CHILD_1" and its nested child window "SUB_CHILD_1"?

  

```

        ImGui::Begin("Main Window");


        //------------------------------------CHILD_1-----------------------------------------------------------
        ImGui::PushStyleColor(ImGuiCol_ChildBg, IM_COL32(255, 0, 255, 255));
        ImGui::SetCursorPos({ 0,0 });
        if (ImGui::BeginChild("CHILD_1", { 400,400 }, ImGuiChildFlags_Border, ImGuiWindowFlags_None) == true)
        {
            ImGui::SetCursorPos({ 0,380 });
            ImGui::Text("CHILD_1");

            ImGui::PushStyleColor(ImGuiCol_ChildBg, IM_COL32(0, 255, 0, 255));
            ImGui::SetCursorPos({ 0,0 });
            if (ImGui::BeginChild("SUB_CHILD_1", { 300,300 }, ImGuiChildFlags_Border, ImGuiWindowFlags_None) == true)
            {
                ImGui::SetCursorPos({ 0,280 });
                ImGui::Text("SUB_CHILD_1");
            }

            ImGui::EndChild(); // Закрываем дочернее окно
            ImGui::PopStyleColor();

        }

        ImGui::EndChild(); // Закрываем дочернее окно

        ImGui::PopStyleColor();
        //-------------------------------------CHILD_1----------------------------------------------------------




        //-------------------------------------CHILD_3----------------------------------------------------------
        ImGui::PushStyleColor(ImGuiCol_ChildBg, IM_COL32(100, 0, 100, 255));

        ImGui::SetCursorPos({ 50,500 });
        if (ImGui::BeginChild("CHILD_3", { 400,400 }, ImGuiChildFlags_Border, ImGuiWindowFlags_None) == true)
        {
            ImGui::SetCursorPos({ 0,380 });
            ImGui::Text("CHILD_3");
        }

        ImGui::EndChild(); // Закрываем дочернее окно

        ImGui::PopStyleColor();
        //--------------------------------------CHILD_3---------------------------------------------------------



        //--------------------------------------CHILD_2---------------------------------------------------------
        ImGui::PushStyleColor(ImGuiCol_ChildBg, IM_COL32(0, 0, 255, 255));


        ImGui::SetCursorPos({ 200,50 });
        if (ImGui::BeginChild("CHILD_2", { 400,400 }, ImGuiChildFlags_Border, ImGuiWindowFlags_None) == true)
        {
            ImGui::SetCursorPos({ 0,380 });
            ImGui::Text("CHILD_2");
        }

        ImGui::EndChild(); // Закрываем дочернее окно

        ImGui::PopStyleColor();
        //--------------------------------------CHILD_2---------------------------------------------------------


        ImGui::End();
        
        
```

![image](https://github.com/user-attachments/assets/1ea789a0-1750-4704-889c-3c18ad552da0)

**Comment by wyxather** [comment]
while reading this code: https://github.com/ocornut/imgui/blob/master/imgui.cpp#L5062
possibly because child2 use main window drawlist and we are rendering the main window first,  if we overlap child_1, child_2, child_3, the order should be child_1(back),child_3,child_2(front), then we go to each child of the window (child_1) and add that to draw data which automatically makes it on top the previous child_2

**Comment by ocornut** [comment]
Child windows are drawn by the parent window in the order they were submitted for a given depth.

It is not expected that you use overlapping child as I don't see a reason for doing so, so this isn't going to be supported.

(Somehow related #4493, #2658)

As this point it is quite hard for me to understand how you are using dear imgui, because you are stumbling on many fringe issues for use cases that all seems rather unusable. Actual context of what you are trying to achieve would be useful sometimes.

**Comment by ElectroidDes** [comment]
> As this point it is quite hard for me to understand how you are using dear imgui, because you are stumbling on many fringe issues for use cases that all seems rather unusable. Actual context of what you are trying to achieve would be useful sometimes.

Yes, I understand more and more that Imgui is poorly designed and written and is not applicable for real use, but is suitable only for simple interfaces for cheat application creators.

**Comment by ocornut** [comment]
You are not answering my request for actual context. If you are asking so entitled please use another software.

**Comment by GamingMinds-DanielC** [comment]
> Yes, I understand more and more that Imgui is poorly designed and written and is not applicable for real use, but is suitable only for simple interfaces for cheat application creators.

In my experience ImGui is suited for very real and complex applications. With a few tricks here and there it can be used for unusual things as well without much of a problem, if you know what you're doing that is.

I have a few decades of programming experience, but so far I can't think of any useful use case that would benefit from an overlapping setup like yours. Since you are talking about "real use", maybe I'm just missing something here. Please enlighten us with your expert knowledge. I mean, with such a cocky and entitled behavior you surely must be able to back it up, right? Right?

**Comment by ocornut** [comment]
It doesn't matter how much programming experience one does have. We need actual context of their real usage to come with best answers.

**Comment by ElectroidDes** [comment]
> Не имеет значения, какой у вас опыт программирования. Нам нужен реальный контекст их реального использования, чтобы получить наилучшие ответы.

Moving children via the mouse cursor, just like it happens with the main window.
But when moving a child window behind the mouse cursor over another child window - the described problem appears.

**Comment by ocornut** [comment]
We need more details on your use case.

If you want movable windows generally you would use child windows.

**Comment by ElectroidDes** [comment]
> Нам нужна более подробная информация о вашем сценарии использования.
> 
> Если вам нужны подвижные окна в целом, вы должны использовать дочерние окна.

I use Child windows, I described everything above. I move the child window CHILD_2, above CHILD_1 - but a Bug occurs, the child window SUB_CHILD_1 - which is drawn BEFORE the child window CHILD_2, which I move - OVERLAPS the child window CHILD_2.

Or does Imgui already implement child windows that can be moved by mouse cursor by default?

**Comment by ocornut** [comment]
We need to understand in which context and why you need child windows to be moved with the mouse, in order to help you find a solution. Otherwise this is possibly a [XY Problem](https://xyproblem.info/).

**Comment by ElectroidDes** [comment]
> Нам нужно понять, в каком контексте и зачем нужно перемещать дочерние окна с помощью мыши, чтобы помочь вам найти решение. В противном случае это, возможно, [проблема XY](https://xyproblem.info/).



> Imgui is poorly designed and written and is not applicable for real use, but is suitable only for simple interfaces for cheat application creators.

What other context?
"why you need child windows to be moved with the mouse" Is this a joke??
I wonder why you made the main Imgui window move with the mouse?

This is exactly what I wrote above: Imgui is poorly designed and written and is not applicable for real use, but is suitable only for simple interfaces for cheat application creators.

In a properly designed and written library - such mildly "strange" questions like "why you need child windows to be moved " should not arise at all. Don't you agree?

**Comment by ocornut** [comment]
I don't agree. If you had read Issue Guidelines you would understand that context always matters and strange requests are frequently XY Problems. I am blocking you now, you are wasting our time. Please seek support elsewhere.

**Comment by ElectroidDes1** [comment]
> I don't agree. If you had read Issue Guidelines you would understand that context always matters and strange requests are frequently XY Problems. I am blocking you now, you are wasting our time. Please seek support elsewhere.

I expected such an answer - it is typical for those who are bad at writing libraries that are only good for simple use. You know very well what many programmers think about Imgui in the context of at least moderately complex interfaces, not to mention complex ones. That's why Imgui is only good for simple things like cheat codes.

Moreover, I didn't ask for any help - I just wanted an answer that Imgui is not capable of doing the thing I described above and I got this answer, thank you. Which only once again confirmed my opinion and that of many other programmers.

**Comment by ElectroidDes1** [comment]
> I have a few decades of programming experience

“There is no fool - like an old fool” (c)

**Comment by GamingMinds-DanielC** [comment]
> It doesn't matter how much programming experience one does have. We need actual context of their real usage to come with best answers.

I know. But that was in response to his ramblings about "not applicable for real use". That statement can easily be countered with real use experience.

> I use Child windows,

When Omar wrote "If you want movable windows generally you would use child windows." he obviously meant "wouldn't" (just a typo, can happen to anyone). After all, that is the conceptual difference between normal windows and child windows. Normal windows are moveable by the user, child windows are usually fixed inside their parent. You are using the wrong tool for the job and are blaming the tool, in a very rude and entitled way on top.

**Comment by ElectroidDes11** [comment]
> > It doesn't matter how much programming experience one does have. We need actual context of their real usage to come with best answers.

This is exactly what I wrote above: Imgui is poorly designed and written and is not applicable for real use, but is suitable only for simple interfaces for cheat application creators.

> I know. But that was in response to his ramblings about "not applicable for real use". That statement can easily be countered with real use experience.

But of course you won't do that, all you'll do is just write that it can be refuted and that's it, without any real refutations.

I mean, with such a cocky and entitled behavior you surely must be able to back it up, right? Right?

> When Omar wrote "If you want movable windows generally you would use child windows." he obviously meant "wouldn't" (just a typo, can happen to anyone)

The typo is not corrected, so it's your wet fantasy.

> Not counting, but an old fool that helped you in previous issues with problems you couldn't solve yourself.

I assumed you would write this - it is typical for those who think that the amount of time programming is somehow related to their experience and abilities. If you have gray hair or are old - it does not mean that you have become smarter.
“There is no fool like - an old fool” (c)


Most programmers know that Imgui is a simple dumb library, which is intended only for simple things. The author of the library is an absolutely bad programmer, for 10 years he was not able to do something normal and you know it and that is why you, like a cowardly woman, ban for such messages and will continue to ban - because you are a coward ;)

**Comment by GamingMinds-DanielC** [comment]
Your ramblings sound like you would blame a toaster for being bad at making coffee.

> But of course you won't do that, all you'll do is just write that it can be refuted and that's it, without any real refutations.
> 
> I mean, with such a cocky and entitled behavior you surely must be able to back it up, right? Right?

How original, a "No, you!" comeback. But sure I can. For tons upon tons of real use examples just look at the gallery threads:
https://github.com/ocornut/imgui/labels/gallery

If you scroll through those threads, you will see a lot of people doing things with this library that you claim this library can't do. Still waiting for an example of an actual use case for your construct.

**Comment by ElectroidDes11** [comment]
> How original, a "No, you!" comeback. But sure I can. For tons upon tons of real use examples just look at the gallery threads: [ gallery](https://github.com/ocornut/imgui/labels/gallery)

It's all a simple interface - there's nothing complicated there. When it comes to at least medium complexity, not to mention high complexity, Imgui sucks, and Omar immediately hides under the table, saying that it's "an XY problem" )))))

> Still waiting for an example of an actual use case for your construct.
Are you a blind stupid old man?
I wrote above the application of this. Moreover, you, a fool, should not concern yourself with how and why I or my users will use this.
There is only one fact Imgui - it can not do this or can but with a bug - потому что Omar не может нормально спроектировать код и написать его. That's all you need to know.
